gammapy/cube/sherpa_.py

```diff
@@ -235,16 +290,118 @@ def __init__(self, name='cube-model', use_psf=True, exposure=None, psf=None, spa
 
             _spatial = result_convol.ravel()
+            spectral_1d = self.spectral_model.calc(pars[self._spectral_pars], ee_lo, ee_hi)
+            _spectral = (spectral_1d.reshape(len(ee_lo), 1, 1) * np.ones_like(xx_lo)).ravel()
         else:
-            _spatial = self.spatial_model.calc(pars[self._spatial_pars], x, y)
-        _spectral = self.spectral_model.calc(pars[self._spectral_pars], elo, ehi)
+            _spatial = self.spatial_model.calc(pars[self._spatial_pars], xlo, xhi, ylo, yhi)
+            _spectral = self.spectral_model.calc(pars[self._spectral_pars], elo, ehi)
         return _spatial * _spectral
+
+
+class CombinedModel3DIntConvolveEdisp(ArithmeticModel):
+    """
+    Combined spatial and spectral 3D model taking into account the energy resolution
+     with the possibility to convolve the spatial model*exposure by the PSF.
+
+    Parameters
+    ----------
+    dimensions: tuple
+        tuple containing the dimension for the images (x,y) and for the reco and true energies. It has to be like
+        [dim_x,dim_y,dim_Ereco,dim_Etrue]
+    use_psf: bool
+        if true will convolve the spatial model by the psf
+    exposure: `~gammapy.cube.SkyCube`
+        Exposure Cube
+    psf: `~gammapy.cube.SkyCube`
+        Psf cube
+    spatial_model: `~sherpa.models`
+        spatial sherpa model
+    spectral_model: `~sherpa.models`
+        spectral sherpa model
+    edisp: `~numpy.array`
+        2D array in (Ereco,Etrue) for the energy dispersion
+    """
+
+    def __init__(self, dimensions, name='cube-model', use_psf=True, exposure=None, psf=None, spatial_model=None,
+                 spectral_model=None, edisp=None):
+        self.spatial_model = spatial_model
+        self.spectral_model = spectral_model
+        self.dim_x, self.dim_y, self.dim_Ereco, self.dim_Etrue = dimensions
+        self.use_psf = use_psf
+        self.exposure = exposure
+        self.psf = psf
+        self.edisp = edisp
+        self.true_energy = EnergyBounds(self.exposure.energies("edges"))
+
+        # The shape of the counts cube in (Ereco,x,y)
+        self.shape_data = (self.dim_Ereco, self.dim_x, self.dim_y)
+        # Array that will store the result after multipliying by the energy resolution in (x,y,Etrue,Ereco)
+        self.convolve_edisp = np.zeros(
+            (self.dim_x, self.dim_y, self.dim_Etrue, self.dim_Ereco))
+
+        # Fix spectral ampl parameter
+        spectral_model.ampl = 1
+        spectral_model.ampl.freeze()
+
+        pars = []
+        for _ in spatial_model.pars + spectral_model.pars:
+            setattr(self, _.name, _)
+            pars.append(_)
+
+        self._spatial_pars = slice(0, len(spatial_model.pars))
+        self._spectral_pars = slice(len(spatial_model.pars), len(pars))
+        ArithmeticModel.__init__(self, name, pars)
+
+    def calc(self, pars, elo, xlo, ylo, ehi, xhi, yhi):
+        from scipy import signal
```

Comment:
I don't think this is  the right place for the import statement, because it's executed in every iteration of the fit, can you move to `__init__()` and store the fftconvolve as e.g. `self._fftconvolve`